icx-proxy/src/main.rs

```diff
@@ -204,10 +210,28 @@ async fn forward_request(
     }
 
     let canister = HttpRequestCanister::create(agent.as_ref(), canister_id.clone());
-    let result = canister
-        .http_request(method, uri.to_string(), headers, &entire_body)
-        .call()
-        .await;
+    let result = match method.to_uppercase().as_str() {
+        "DELETE" | "PATCH" | "POST" | "PUT" => {
+            // We need to fetch the root key for updates.
+            if fetch_root_key {
+                agent.fetch_root_key().await?;
+            }
```

Comment:
`fetch_root_key` actually always calls the status endpoint and sets the root key.  I'd recommend to call it once right after creating the `Agent`, if `--fetch-root-key` was passed, rather than on every request of one of these types.